Refactor ValidatorSet

AndrejMitrovic opened the pull request:
See each commit. I put a lot of effort into making the commits as atomic as possible.

Part of #836

AndrejMitrovic committed:
Refactor ValidatorSet's usage of SQLite

Don't store serialized Enrollment blobs anymore.
Instead store the key, cycle length, enrolled height,
and the current distance and associated preimage.

This enables SQL queries which look up preimages
with the given distance.

The signature is discarded as it's no longer needed
after validation.

Removed many unnecessary (de)serializations.

Replaced private getEnrollment() function with
getOlderPreimageDistance() which can find
a preimage with an older distance and a matching
UTXO key.

Part of #836

AndrejMitrovic committed:
Change distance type ulong => ushort

The distance will never exceed ValidatorCycle - 1,
which is 1007 and can fit within 2 bytes (ushort).

AndrejMitrovic committed:
Replace getValidators => getEnrolledUTXOs

We do not need the Enrollments, we only
need the UTXO keys from the database.

AndrejMitrovic committed:
QuorumGenerator: Use UTXO keys instead of Enrollments

We don't need the Enrollment structure, we only need
the UTXO keys.

In one of the next commits the EnrollmentManager's
getValidators() function will be replaced with
one which just returns the UTXO keys.

Splitting this into multiple commits to make it atomic.

AndrejMitrovic committed:
getPreimage: Replace bool return with PreImageInfo

Previously this function was inconsistently returning
true / false, returning true when the preimage was
missing, and only returning false if there was a
database error.

Database errors cannot (and were not) handled by client code,
while it was already handling the PreImageInfo.init case.
Therefore the function can simply return the PreImageInfo
directly.

Fixes #942

AndrejMitrovic committed:
ValidatorSet: Make all methods nothrow

The API was inconsistent and treated some
calls as nothrow while others not.

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -436,86 +406,88 @@ public class ValidatorSet
     ***************************************************************************/
 
-    public bool addPreimage (const ref PreImageInfo preimage) @safe nothrow
+    private bool getOlderPreimageDistance (Hash enroll_hash, ushort new_distance,
+        out Hash preimage, out ushort distance) @trusted nothrow
```
You can merge this with `getPreimage` if you change the signature of `getPreimage` to:
```D
public PreImageInfo getPreimage (const ref Hash enroll_key, ushort min_distance = 0) @trusted nothrow
```

AndrejMitrovic replied:
Haha I made the same comment just a minute before you. :)

AndrejMitrovic commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -436,86 +406,88 @@ public class ValidatorSet
+    {
+        Hash prev_preimage;
+        ushort prev_distance;
+        if (!this.getOlderPreimageDistance(preimage.enroll_key,
```
Maybe we don't need this. Maybe we could just add a `new_distance` parameter to `getPreimage`. It would simplify the implementation.

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -436,86 +406,88 @@ public class ValidatorSet
+                    preimage.hash.toString(), preimage.distance,
+                    preimage.enroll_key.toString());
             }();
         }
```
Instead of doing a `SELECT` then `UPDATE`, do `UPDATE` + `WHERE` ?

AndrejMitrovic replied:
Not a bad idea.

AndrejMitrovic replied:
Although, how would I verify the preimage hash matches? I still need to read the old distance and then call `isInvalidReason`.

Geod24 replied:
Damn, right.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -453,9 +453,7 @@ public class FullNode : API
-        PreImageInfo preimage;
-        this.enroll_man.getValidatorPreimage(enroll_key, preimage);
-        return preimage;
+        return this.enroll_man.getValidatorPreimage(enroll_key);
```
Technically here we should throw an Exception for the generator to pick up a non-200 is needed. But that works too.

AndrejMitrovic replied:
Hmm. It seems then we need to be more consistent. `getEnrollment` / `getMerklePath` also return `.init`. Should we file an issue for this?

Geod24 replied:
Yes

Geod24 commented on source/agora/consensus/EnrollmentManager.d:
```diff
@@ -398,22 +398,22 @@ public class EnrollmentManager
     /***************************************************************************
 
-        Get validator's pre-image from the validator set
+        Get validator's pre-image from the validator set.
```
Commit message gas a typo (`PreImangeInfo`)

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -103,16 +104,16 @@ public class ValidatorSet
-            serializeToBuffer(img, pbuffer);
-
             () @trusted {
+                // todo: we probably don't need to store enrolled_height here
+                // as it's only used for *this* validator
```
Can you expand on this ?  Enrolled height is used to tell when we need to remove a validator, isn't it ?

AndrejMitrovic replied:
Ah right, brain slip. I'll remove this.

Geod24 commented on source/agora/consensus/ValidatorSet.d:
```diff
@@ -34,6 +34,8 @@ import d2sqlite3.library;
 import d2sqlite3.results;
 import d2sqlite3.sqlite3;
 
+import ocean.text.convert.Formatter;
```
Can we just fix the bindings instead ?

AndrejMitrovic replied:
Alright. I'll remove the last commit for now.

Geod24 reviewed (commented):
Broadly LGTM, except for the "Remove utxo key comparison in PreImage validation" commit. I think our validation routines should be self-consistent, and assuming a prior check is a bit odd to me.

AndrejMitrovic commented:
I'll push the changes tomorrow, I have a commit that's still on my laptop.

AndrejMitrovic commented:
Updated.

I haven't added `min_distance` to `getPreimage` because what we actually want is `max_distance` (we're specifically looking for an older preimage). And if I used `max_distance` it would be an awkward API because we already have `hasPreimage` that asks for a minimum distance. The two APIs would have opposite meanings of the second parameter.


However it was a micro-optimization anyway, we don't really need it.

AndrejMitrovic commented:
Oh but I did remove `getOlderPreimageDistance`, since I also removed the `Remove utxo key comparison in PreImage validation` commit. Also removed the string buffer optimization.
